Feature/domain core types

AlexBud369 opened the pull request:
# Add Role enum and repository interfaces for domain layer
- Added `Role` enum to define user roles (`User`, `Admin`) for role-based authorization
- Added `IUserRepository` for user management operations (CRUD, block, unblock)
- Added `IBookRepository` for book management operations (CRUD, search, filter, pagination)
- Added `ITagRepository` for tag management operations (CRUD)
- Added `IRefreshTokenRepository` for refresh token management operations (CRUD, retrieval by token)

AlexBud369 committed:
feat(domain): add User entity with authentication and book management

- Implemented User entity with fields for id, username, email, password hash, role, and book collection
- Added validation for username, email, and password hash
- Included methods for updating user details, role, and blocking status
- Supported EF Core with private constructor and private setters

AlexBud369 committed:
feat(domain): add Book entity for book management

- Implemented Book entity with fields for id, title, author, description, cover image, and tags
- Added validation for title and author
- Included methods for updating details, managing tags, and checking deletion permissions
- Supported EF Core with private constructor and private setters

AlexBud369 committed:
feat(domain): add RefreshToken entity for authentication

- Implemented RefreshToken entity with fields for id, token, user, creation, expiration, and revocation
- Added constructor with validation for token value
- Included Revoke method for invalidating tokens
- Supported EF Core with private constructor

AlexBud369 committed:
feat(domain): add Tag entity for book categorization

- Implemented Tag entity with fields for id, name and book collection
- Added private constructor for EF Core support and encapsulation
- Included validation through private setter for TagName
- Added UpdateTagName method for controlled modification

AlexBud369 committed:
fix: refactor Book entity
- replaced private fields with auto-properties
- removed methods UpdateDetails, UpdateCoverImageUrl, AddTag, RemoveTag, CanBeDeletedBy
-removed ValidateNotEmpty from constructor.

AlexBud369 committed:
fix: refactor User entity
- replaced private fields with auto-properties
- removed methods UpdateName, UpdateEmail, UpdatePasswordHash, ChangeRole, Block, Unblock, AddBook, AddRefreshToken
- removed ValidateNotEmpty from constructor
- used IReadOnlyList<T> for RefreshTokens and Books with private backing Lists
- fixed redundant UserId initialization in constructor

AlexBud369 committed:
fix: refactor Tag entity
- replaced private fields with auto-properties
- removed UpdateTagName method
- removed redundant tagId parameter from constructor
- used IReadOnlyList<Book> for Books with private backing List
- removed validation from constructor

AlexBud369 committed:
fix: refactor RefreshToken entity
- converted to file-scoped namespace
- replaced private fields with auto-properties
- remove validation in constructor

AlexBud369 committed:
feat(domain): add Enums folder with UserRole enum

AlexBud369 committed:
feat: add IUserRepository interface for user data operations (CRUD, block, unblock)

AlexBud369 committed:
feat: add IBookRepository interface for book data operations (CRUD, search, filter, pagination)

AlexBud369 committed:
feat: add IRefreshTokenRepository interface for refresh token operations

AlexBud369 committed:
feat: add ITagRepository interface for tag data operations

AlexBud369 committed:
fix: remove Domain project ref from Infrastructure layer

budnikme commented on src/Domain/Enums/Role.cs:
```diff
@@ -0,0 +1,8 @@
+
+public enum Role
+{
+    User,
```
Please specify enum values explicitly to avoid potential bugs in the feature

```
public enum Role {
    User = 1,
    Admin = 2
}
```

budnikme commented on src/Domain/Repositories/IUserRepository.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace Domain.Repositories;
+
+public interface IUserRepository
```
Ef core already implements repository pattern, so there is no sense to implement it by ourselves again

AlexBud369 committed:
fix: relocate repository interfaces to Application layer
- Moved repositories to Application/Interfaces to adhere to Clean Architecture principles
- Removed Domain/Repositories folder

AlexBud369 committed:
fix: explicitly specify Role enum values

AlexBud369 committed:
refactor: remove repository interfaces

AlexBud369 committed:
fix: remove public constructors from entities
- Removed public constructors from entities
- Kept private parameterless constructors for EF Core compatibility
- Changed mutable properties to use 'set' instead of 'private set'

AlexBud369 committed:
fix: remove RefreshToken entity and Role enum
- removed RefreshToken entity as ASP.NET Core Identity will handle refresh tokens
- removed Role enum in preparation for using IdentityRole<Guid>
- updated User entity to remove RefreshTokens and Role properties temporarily

AlexBud369 committed:
fix: Replace User with ApplicationUser for ASP.NET Core Identity
- Removed redundant fields (userId, Username, Email, PasswordHash) provided by IdentityUser
- Kept IsBlocked and Books properties for custom functionality

AlexBud369 committed:
feat: add ApplicationRole for ASP.NET Core Identity
- Added ApplicationRole class inheriting from IdentityRole<Guid> to support role-based authorization
- Included private parameterless constructor for EF Core compatibility

AlexBud369 committed:
chore: add Microsoft.AspNetCore.Identity.EntityFrameworkCore package

AlexBud369 committed:
fix: remove private parameterless constructors from entities

AlexBud369 committed:
feat: restore RefreshToken entity and update related entities
- Restored RefreshToken entity to support refresh token functionality with ASP.NET Core Identity
- Added RefreshTokens property to ApplicationUser for one-to-many relationship
- Renamed tagId to TagId and tagName to TagName in Tag entity for consistent naming

AlexBud369 commented:
### Key Changes  

#### 1. **ASP.NET Core Identity Integration**  
- Replaced custom `User` entity with `ApplicationUser`  
- Added `ApplicationRole` for role management  

#### 2. **Entity Refactoring**  
- Modified constructors (private for EF Core, removed public ones)  
- Made entity properties mutable (changed `private set` to `set`)  
- Restored `RefreshToken` entity with proper relationships  

#### 3. **Architecture Adjustments**  
- Moved repository interfaces to `Application/Interfaces` (Clean Architecture compliance)  
- Removed redundant `Domain/Repositories` folder  

#### 4. **Role Management**  
- Removed legacy `Role` enum  
- Transitioned to `IdentityRole<Guid>` for built-in role handling

AlexBud369 committed:
feat: add DomainConstants file to manage constants

AlexBud369 committed:
feat: update DomainConstants with user roles and refresh token
